modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordExecManager.kt

```diff
@@ -79,44 +68,129 @@ object MirrordExecManager {
         return path
     }
 
-    fun start(wslDistribution: WSLDistribution?, project: Project, product: String): Map<String, String>? {
-        return start(wslDistribution, project, null, product)?.first
+    private fun getConfigPath(configFromEnv: String?): String? {
+        return if (ApplicationManager.getApplication().isDispatchThread) {
+            // We can create a file via VFS from the dispatch thread.
+            WriteAction.compute<String, InvalidProjectException> {
+                service.configApi.getConfigPath(configFromEnv, true)
+            }
+        } else if (ApplicationManager.getApplication().isReadAccessAllowed) {
+            // This thread already holds a read lock, and it's not a dispatch thread.
+            // There is no way to create a file via VFS here.
+            // We abort if the config does not yet exist.
+
+            service.configApi.getConfigPath(configFromEnv, false)?.let {
+                return it
+            }
+
+            service
+                .notifier
+                .notification("mirrord requires configuration", NotificationType.WARNING)
+                .withAction("Create default") { e, n ->
+                    e
+                        .project
+                        ?.service<MirrordProjectService>()
+                        ?.let {
+                            val config = WriteAction.compute<VirtualFile, InvalidProjectException> {
+                                service.configApi.createDefaultConfig()
+                            }
+                            FileEditorManager.getInstance(service.project).openFile(config, true)
+                        }
+                    n.expire()
+                }
+                .fire()
+            null
+        } else {
+            // This thread does not hold any lock,
+            // we can safely wait for the config to be created in the dispatch thread.
+            var config: Pair<String?, InvalidProjectException?> = Pair(null, null)
+
+            ApplicationManager.getApplication().invokeAndWait {
+                config = try {
+                    val path = service.configApi.getConfigPath(configFromEnv, true)
+                    Pair(path, null)
+                } catch (e: InvalidProjectException) {
+                    Pair(null, e)
+                }
+            }
+
+            config.second?.let { throw it }
+            config.first!!
```

Comment:
I am not sure I understand what this does? also we should avoid `!!`, will cause exceptions

Reply:
Above that I run some code in the EDT thread using `invokeAndWait`. That code may throw an exception. If it's not caught, it produces an ugly error notification in the UI. Therefore, I catch the exception and pass the result/exception back to the background thread wrapped in the `Pair<String, Exception>`. One of those is always not null, so tbh I think `!!` is ok

Reply:
can we handle this null case by any chance and propagate the error back to the user?

Reply:
I got rid of `!!` and moved the notification out of this method